### PR TITLE
Support implicit type conversions

### PR DESCRIPTION
**Pull request by fiftin**
Relates to https://github.com/ClickHouse/clickHouse-go/issues/574

Currently, v2 has a number of functional gaps where no precision would be lost and conversion could be safely performed. This PR adding following implicit conversions:

* UUID <-> string
* IPv4 <-> string
* IPv6 <-> string

### CODE REVIEW

File: lib/column/decimal.go [hunk 1]
```diff
@@ -92,6 +92,11 @@ func (col *Decimal) Row(i int, ptr bool) interface{} {
 
 func (col *Decimal) ScanRow(dest interface{}, row int) error {
 	switch d := dest.(type) {
+	case *float64:
```
Comment (gingerwizard) [inline lib/column/decimal.go:95]: This potentially results in precision loss e.g. float64 can't represent numbers exactly like decimal. This is against a philosophy we apply in v2 that I think v1 got wrong i.e. it allowed precision loss. The user should accept this in their code explicitly vs the driver doing it for them.

**Comment by gingerwizard** [comment]
@fiftin thanks for this. Comfortable with the ip/uuid conversions- they seem useful and we can reuse them elsewhere (@genzgd thinking JSON on read). Not sure on whether allowing float64 and Decimal to be interchangeable is wise - since the former can't represent all values of the latter - see the comment.

File: lib/column/decimal.go (hunk 1 above)
Comment (fiftin) [reply lib/column/decimal.go:95]: Ok, what about IP and UUID?

File: lib/column/ipv4_test.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+	// appending string
+	err := col.AppendRow(invalidIPv4Str)
+	if err == nil {
+		t.Fatalf("AppendRow didn't IP")
+	}
+	switch e := err.(type) {
+	case *ColumnConverterError:
+		if e.Hint != "invalid IP format" {
+			t.Fatalf("expected error hint %s, got %s", "AppendRow didn't IP", e.Hint)
```
Comment (gingerwizard) [inline lib/column/ipv4_test.go:21]: can we use `require` instead and avoid indentation? See other tests. e.g. https://github.com/ClickHouse/clickhouse-go/blob/v2/tests/json_test.go#L269

File: lib/column/ipv4_test.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+	// appending strings
+	col := IPv4{}
+	err := col.AppendRow("127.0.0.1") // add 1 valid IP
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = col.Append(strIps)
+	if err == nil {
+		t.Fatal("Append must return error if invalid IP provided")
```
Comment (gingerwizard) [inline lib/column/ipv4_test.go:43]: see above using require and assert packages in test.

File: lib/column/ipv4.go [hunk 4]
```diff
@@ -51,6 +51,11 @@ func (col *IPv4) Row(i int, ptr bool) interface{} {
 
 func (col *IPv4) ScanRow(dest interface{}, row int) error {
 	switch d := dest.(type) {
+	case *string:
+		*d = col.row(row).String()
+	case **string:
```
Comment (gingerwizard) [inline lib/column/ipv4.go:56]: where is the need for support for this coming from? can you give me an example as to when we need  **string support?

File: lib/column/ipv4_test.go (hunk 2 above)
Comment (fiftin) [reply lib/column/ipv4_test.go]: Done

File: lib/column/ipv4_test.go (hunk 3 above)
Comment (fiftin) [reply lib/column/ipv4_test.go]: Done

File: lib/column/ipv4.go (hunk 4 above)
Comment (fiftin) [reply lib/column/ipv4.go:56]: I relied on String column's ScanRow method.

https://github.com/ClickHouse/clickhouse-go/blob/a27ceead7ef92930e47841cec6d08cc24912c4dd/lib/column/string.go#L61

File: lib/column/decimal.go (hunk 1 above)
Comment (fiftin) [reply lib/column/decimal.go]: Removed

**Comment by fiftin** [comment]
@gingerwizard ,

What do you think about Geo types conversion to GeoJSON?

File: lib/column/ipv4.go (hunk 4 above)
Comment (gingerwizard) [reply lib/column/ipv4.go:56]: yes i noticed this, let me see if that's actually something useful..i might be missing a use case.

File: lib/column/ipv4.go [hunk 5]
```diff
@@ -66,39 +71,93 @@ func (col *IPv4) ScanRow(dest interface{}, row int) error {
...
+			Op:   "Append",
+			To:   "IPv4",
+			From: "IPv6",
+			Hint: "invalid IP version",
+		}
+	}
+	return append(data, IPv4ToBytes(ip)...), nil
+}
+
+func (col *IPv4) appendIPv4(ip net.IP) (err error) {
```
Comment (gingerwizard) [inline lib/column/ipv4.go:103]: Do we need these one liner methods?  Not sure they really refactor a huge amount and just make the code harder to follow

Comment (fiftin) [reply lib/column/ipv4.go]: Removed

**Comment by fiftin** [comment]
All done, please check.

**Pull request merged**